experimental/lib/Analysis/GsaAnalysis.cpp

```diff
@@ -0,0 +1,485 @@
+// assignment (SSA) representation into gated single assignment representation
+// (GSA).
+//
+//===----------------------------------------------------------------------===//
+
+#include "experimental/Analysis/GsaAnalysis.h"
+#include "dynamatic/Dialect/Handshake/HandshakeOps.h"
+#include "experimental/Support/BooleanLogic/BDD.h"
+#include "experimental/Support/FtdSupport.h"
+#include "mlir/Analysis/CFGLoopInfo.h"
+#include "mlir/IR/Dominance.h"
+#include "vector"
+#include "llvm/Support/Debug.h"
+#include <algorithm>
+
+using namespace mlir;
+using namespace mlir::func;
+using namespace dynamatic::experimental::ftd;
+
+namespace dynamatic {
+namespace experimental {
+namespace gsa {
+
+template <typename FunctionType>
+Gate *
+GsaAnalysis<FunctionType>::expandGammaTree(ListExpressionsPerGate &expressions,
+                                           std::vector<std::string> &conditions,
+                                           Gate *originalPhi) {
+
+  // At each iteration, we want to use a cofactor that is present in all the
+  // expressions in `expressions`. Since the cofactors are ordered according to
+  // the basic block number, we can say that if a cofactor is present in one
+  // expression, then it must be present in all the others, since they all have
+  // the blocks associated to that cofactor as common dominator.
+  std::string conditionToUse;
+  while (true) {
+    conditionToUse = conditions.front();
+    conditions.erase(conditions.begin());
+    unsigned cofactorUsage =
+        std::count_if(expressions.begin(), expressions.end(),
+                      [&](std::pair<boolean::BoolExpression *, GateInput *> g) {
+                        return g.first->containsMintern(conditionToUse);
+                      });
+
+    // None is using that cofactor, go to the next one
+    if (cofactorUsage == 0)
+      continue;
+
+    // All are using the cofactor, expand the expressions over it
+    if (cofactorUsage == expressions.size())
+      break;
+
+    assert(false && "A cofactor is used only by some expressions");
+  }
+
+  ListExpressionsPerGate conditionsFalseExpressions;
+  ListExpressionsPerGate conditionsTrueExpressions;
+
+  // For each expression
+  for (auto expression : expressions) {
+
+    // Restrict an expression according to a condition ans an input boolean
+    // value and return it
+    auto getRestrictedExpression =
+        [&](bool value) -> boolean::BoolExpression * {
+      auto *expr = expression.first->deepCopy();
+      boolean::restrict(expr, conditionToUse, value);
+      return expr->boolMinimize();
+    };
+
+    // Substitute a condition in the expression both with value true and false
+    auto *exprTrue = getRestrictedExpression(true);
+    auto *exprFalse = getRestrictedExpression(false);
+
+    // One of the two expressions above will be zero: add the input to the
+    // corresponding list
+    if (exprTrue->toString() != "0")
```

Comment:
I suspect this is not your fault and is the code you inherited, but it is really really bad design that we have to check against a string for some kind of equality (something I pointed out at the time but which was ignored). If there is a small fix that would allow to change that, please do it. Otherwise don't bother.